Improvements to event handling

rynowak opened the pull request:
The changes here make event dispatching (including bind) more
user-friendly by avoiding the need for manual calls to StateHasChnaged.

We also introduce a new type `EventCallback` (and `EventCallback<T>`).
This is a new primitive that is like a super-powered version of a
delegate. When writing a component that accepts delegates as parameters,
consider using `EventCallback` for the following reasons:
- Allows consumer to pass a variety of different delegate signatures
- Does proper event dispatching and error handling

Using `EventCallback` will eliminate most of the remaining cases where a
manual `StateHasChanged` is required when components are passing content
and delegates to each other.

`EventCallback` is inherently async for the reason that this is really
the only way to provide correct error handling.

-----

The fix for this will be two-phase by first creating a set of APIs that
can be targeted by the compiler that has the desired behaviour and then
updating the compiler to target this new infrastructure.

rynowak commented on src/Components/Components/src/EventHandlerInvokerFactory.cs:
```diff
@@ -0,0 +1,102 @@
+
+namespace Microsoft.AspNetCore.Components
+{
+    public sealed class EventHandlerInvokerFactory
```
Still very much a work in progress, but this should give you the gist of it. 

This is an instance (not a static) so it can be targeted by extension methods.

rynowak commented on src/Components/Components/src/EventHandlerInvokerFactory.cs:
```diff
@@ -0,0 +1,102 @@
+                return (Action)(() =>
+                {
+                    action();
+                    _ = handler.HandleStateChangeAsync(Task.CompletedTask);
```
Note also that this is kinda fire-and-forget. I'm not sure it makes much sense for this method to return a task, because in many many cases we won't be able to ever await it.

javiercn replied:
I think all the above methods need to have Func<Task> based overloads so that we can bind async handlers and flow the task all the way up to the renderer decide what to do

rynowak replied:
Yes, that will be done as well.

rynowak replied:
@javiercn - note that in come cases these event handlers are fire-and-forget. Nothing awaits them and nothing can.

javiercn commented on src/Components/Components/src/ComponentBase.cs:
```diff
@@ -276,41 +308,6 @@ private async Task ReRenderAsyncTask(Task task)
             }
         }
 
-        private async void ContinueAfterLifecycleTask(Task task)
```
Glad to see this go away

javiercn commented on src/Components/Blazor/Blazor/src/Rendering/WebAssemblyRenderer.cs:
```diff
@@ -32,7 +32,10 @@ public WebAssemblyRenderer(IServiceProvider serviceProvider): base(serviceProvid
-            => DispatchEvent(componentId, eventHandlerId, eventArgs);
+        {
+            // Event dispatching from the browser is fire-and-forget
+            _ = DispatchEventAsync(componentId, eventHandlerId, eventArgs);
```
I expect things like error handling to happen at this layer, so either the Renderer does it (on the base class) or we have some code here to at least log to the console or similar (equivalent to what we used to do)

rynowak replied:
The error handling is inside `DispatchEventAsync`

javiercn commented:
I'm not sure I understand the nature of the change. Is this intended to support something like?
```
<div>
   <Child OnClick="InvokeSomeParentFunction" />
</div>
```
And having StateHasChanged be called on automatically after InvokeSomeParentFunction?

javiercn commented on src/Components/Components/src/IHandleStateChange.cs:
```diff
@@ -0,0 +1,25 @@
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// Interface implemented by components that receive notification of state changes.
```
This feels confusing/wrong programming model wise. As far as I understand, state is internal to the component and the component is responsible for triggering renders when its state changes and it needs to rerender.

javiercn replied:
I think I understand now what you are trying to accomplish here. I think this poses the issue that the component might rerender in many cases even if nothing changes, given that it doesn't have a good way to compare its previous state against the current state, and that might introduce performance penalties.

rynowak replied:
The runtime needs to notify the component **somehow** that an event happened. This is the somehow - it's up to the component whether it wants to rerender or not. By default the answer is yes.

SteveSandersonMS replied:
I do agree with @javiercn, though @rynowak of course you're right that we need some way to signal that an event happened.

Perhaps we should change the name. Instead of `IHandleStateChange`, something like `IHandleAfterEvent`, since we're not really saying any state has changed, but just that some event has happened.

Actually I'm looking for all the places where we call `HandleStateChangeAsync` with some task other than `Task.CompletedTask`, and there's only one such place: `ComponentBase.cs` L171. That being the case, might it be possible to keep the earlier interface where we expect to be given `EventHandlerInvoker invoker` and `UIEventArgs args`, and then the recipient is the thing responsible for actually calling the `invoker`? If we could do that, it pretty much completely eliminates the risk that this interface gets misused as a way to invoke `StateHasChanged` on arbitrary components outside the intended lifecycle, preserving the encapsulation so that components know they only trigger that themselves (or the system does).

Perhaps there are other cases where we actually have to start with an existing `Task` rather than an `EventHandlerInvoker`. What do you think?

SteveSandersonMS replied:
Oh actually I guess there's always the scenario like in `SetValueHandler` where there's no delegate to pass, but rather we're just supplying notification that an event has already happened. So maybe you can ignore the second half of my suggestions there.

But I do still think `IHandleAfterEvent` might be a better name than `IHandleStateChange`.

rynowak replied:
I'm going to change this to be void and called `IHandleAfterEvent` per feedback.

rynowak replied:
I ended up changing this back to its original signature. The problem is that if we want the component to return a task, it's helpful for the component to be the one to invoke the delegate. Other solutions (unless we made it synchronous) become *really* complicated.

javiercn commented on src/Components/Components/src/ComponentBase.cs:
```diff
@@ -322,15 +319,18 @@ private static void HandleException(Exception ex)
         {
-            var task = binding.Invoke(args);
-            ContinueAfterLifecycleTask(task);
+            var (isAsync, asyncTask) = ProcessLifeCycletask(task);
```
I imagine you are doing this to rerender on async work. You should do just that. Given that you also seem to agree that event handling needs to be task returning, we should just flow the errors out of the component into the renderer where they can be handled in a centralized place

rynowak replied:
I'm not sure I really get what the suggestion is here. But given that we're changing the signature of this method is - I think it's going to be unimportant. Holla back if you still have feedback after I update this.

rynowak commented:
> I'm not sure I understand the nature of the change. Is this intended to support something like?
> 
> ```
> <div>
>    <Child OnClick="InvokeSomeParentFunction" />
> </div>
> ```
> And having StateHasChanged be called on automatically after InvokeSomeParentFunction?

Yes, exactly this.

rynowak commented:
I've rebased this on top of the current bits and incorporated some of the things from @pranavkm 's work so that we're not fighting each other. I assume his thing is going to go in first. 

My next steps are to incorporate feedback from before London and then make this more complete.

rynowak commented on src/Components/Components/src/EventHandlerInvoker.cs:
```diff
@@ -11,24 +11,30 @@ namespace Microsoft.AspNetCore.Components
     public readonly struct EventHandlerInvoker
     {
-        private readonly MulticastDelegate _delegate;
+        public static readonly EventHandlerInvokerFactory Factory = new EventHandlerInvokerFactory();
```
@SteveSandersonMS  - what would you think about making this a more user-friendly name and making it part of the programming experience. I'm thinking `Command` and `Command<T>` and keeping the ability to abstract over the async-vs-non.

The reason why you might want to have something like this is to deal with the problem when a component wants to use delegates as parameters. The consumer of the component has to provide delegates with the signature that the component specifies. This is a little clunky in comparison to how event-handlers work which accept a variety of signatures seamlessly. This would be done in such a way that the user of a component doesn't need to know about the details of command, a super-set of what we have today would work. 

The other detail is that if a component accepts a `Command` as a parameter, it allows us to back in the same eventing behaviour that we have with event handlers and bind but without guesswork and conventions.

SteveSandersonMS replied:
This sounds superb. I was just hitting cases today with the forthcoming `EditForm` component where it's awkward to have an `OnSubmit` parameter because sometimes the consumer might legitimately want to supply any of `Action`, `Action<EditContext>`, `Func<Task>` or `Func<EditContext, Task>`.

If this could be declared by `EditForm` as `[Parameter] Command<EditContext> OnSubmit { get; set; }` and this makes it possible for the consumer to use any of those delegate types, that would be a huge win.

On naming, the name `Command` might be a good fit, or `EventCallback` is also pretty explict since it comes with the implication that it triggers post-event handling logic.

rynowak replied:
If we want this to support arbitrary T it will require some doing, but shouldn't be that bad. The main that comes up is that it kinda has to be reference type at that point. This might be OK because we can amortize some of the allocations we already have to do - I'm going to have a go at this and we can look at in more detail. 

I want to spend some extra time to really nail this part since I think it nice to have a primitive that we use to events, bind, and arbitrary app functionality like `OnSubmit`.

rynowak commented on src/Components/Components/src/IHandleStateChange.cs:
```diff
@@ -0,0 +1,25 @@
+        /// <returns>
+        /// A <see cref="Task"/> that completes once the component has processed the state change.
+        /// </returns>
+        Task HandleStateChangeAsync(Task task);
```
Per discussion this is going to be made `void HandleStateChange()`

rynowak commented on src/Components/Components/src/Rendering/Renderer.cs:
```diff
@@ -536,6 +545,22 @@ private void RemoveEventHandlerIds(ArrayRange<int> eventHandlerIds, Task afterTa
+                    HandleException(ex);
+                }
+            }
+        }
```
copied from @pranavkm

rynowak commented on src/Components/Components/src/Rendering/Renderer.cs:
```diff
@@ -329,8 +339,7 @@ public virtual Task InvokeAsync(Func<Task> workItem)
             {
-                workItem();
-                return Task.CompletedTask;
+                return workItem();
```
This was a bug that broke an E2E test. I'm not sure how we didn't detect this somehow.

SteveSandersonMS commented on src/Components/Components/src/Rendering/Renderer.cs:
```diff
@@ -267,24 +267,36 @@ private ComponentState AttachAndInitComponent(IComponent component, int parentCo
+                }
+                catch (Exception ex) when (!task.IsCanceled)
+                {
+                    HandleException(ex);
```
It took me a few minutes to realise what scenario L291 is for. Would something like the following make the flow more obvious, or am I misunderstanding the intended logic?

```cs
catch (Exception synchronousException)
{
    task = Task.FromException(synchronousException);
}
```

... so that the flow always continues to L299, and we don't need any extra special case for `IsCanceled`?

SteveSandersonMS replied:
Actually it probably does still need special handling for `IsCanceled` regardless.

rynowak replied:
I based this on @pranavkm 's pr so I don't want to super-diverge from what's going on there. If we think this needs to change that's fine. 

Would comments make a big difference here? I don't relish the thought of allocating an extra task just for funsies.

SteveSandersonMS replied:
Sure, a comment would be totally sufficient.

SteveSandersonMS commented on src/Components/Components/test/RendererTest.cs:
```diff
@@ -436,12 +436,15 @@ public void CanDispatchEventsToTopLevelComponents()
             Assert.Same(eventArgs, receivedArgs);
+            Assert.Equal(TaskStatus.RanToCompletion, task.Status); // Completes synchronously
+
+            await task; // Does not throw
```
This is a great way to express the flow, showing that the effects are synchronous plus the task itself doesn't fail. 👍

SteveSandersonMS commented on src/Components/Components/test/RendererTest.cs:
```diff
@@ -946,7 +1160,7 @@ public void AllRendersTriggeredSynchronouslyDuringEventHandlerAreHandledAsSingle
 
             // Act
-            renderer.DispatchEvent(childComponentId, origEventHandlerId, args: null);
+            await renderer.DispatchEventAsync(childComponentId, origEventHandlerId, eventArgs: null);
```
I think having an `await` here somewhat undermines the idea of this test, which is about what happens synchronously.

Would it be possible *not* to await this task? Or perhaps capture it and put `await capturedTask` at the end of the test.

Ideally it would be preferable to do that for all the tests being updated here, since we explicitly want the effects to remain synchronous.

rynowak replied:
Yeah good catch. I was trying to preserve the semantics of this but I misread the test name. I'll review these again.

SteveSandersonMS commented on src/Components/Shared/test/TestRenderer.cs:
```diff
@@ -43,21 +40,8 @@ public new Task RenderRootComponentAsync(int componentId)
-            }
-        }
+        public new Task DispatchEventAsync(int componentId, int eventHandlerId, UIEventArgs eventArgs)
+            => InvokeAsync(() => base.DispatchEventAsync(componentId, eventHandlerId, eventArgs));
```
Is it necessary to pass the logic through `InvokeAsync` here? I'm concerned this might make the tests describe something different than how the real renderers would behave, since when dispatching events to the real renderers we don't force it through an async hop.

rynowak replied:
Now that we require the use the sync context events have to be dispatched on the sync context. The renderer would already be running on the sync context when `DispatchEventAsync` is called.

So I guess I'm not sure what else this would do? 

Another *very intrusive* approach is a write a test framework that has a sync context, and then just ignore threading in your tests. We have done this a few places, but I don't really relish it here.  When you do this, a different set of test scenarios become hard.

SteveSandersonMS commented on src/Components/Components/src/ComponentBase.cs:
```diff
@@ -157,6 +157,38 @@ protected Task Invoke(Action workItem)
+        /// <returns>
+        /// A <see cref="Task" /> that asynchronously completes once event processing has completed.
+        /// </returns>
+        public static Task DispatchStateChangeAsync(object component, EventHandlerInvoker invoker, UIEventArgs args)
```
Could the public versions of `DispatchStateChangeAsync` take an `IComponent component`, instead of `object component`? Seems like that would make the usage more explicit. I don't think there's any scenario where we expect people to dispatch to something that's not a component.

It's fine if there's also an internal overload that takes `object component` so we don't have to put casts and checks into all the `SetValueHandler` calls.

SteveSandersonMS replied:
Or even `IHandleStateChange` instead of `IComponent`.

SteveSandersonMS replied:
Also the overload that takes `(object component, EventHandlerInvoker invoker, UIEventArgs args)` seems only to be used in one place, and that place could call `invoker.Invoke(args)` itself then use the `(object component, Task task)` overload. So maybe we only need the second overload.

rynowak replied:
Sure. I cribbed this from your original PR - but I think the semantics have changed. You originally had this typed as object or `IComponent` so that callers could at least attempt to dispatch a change manually with minimal ceremony.

Another consideration is whether we want to allow AppState objects to implement `IHandleStateChange`.

rynowak replied:
I think I'd prefer to leave this internal or let it accept object. There are **a lot** of call sites of this, and it doesn't seem useful to expand the boilerplate into all of those locations.

rynowak replied:
This is gone now, so not super relevant.

SteveSandersonMS commented on src/Components/Components/src/BindMethods.cs:
```diff
@@ -74,129 +74,187 @@ public static MulticastDelegate GetEventHandlerValue<T>(Func<T, Task> value)
+            return _ =>
+            {
+                setter((string)((UIChangeEventArgs)_).Value);
+                ComponentBase.DispatchStateChangeAsync(setter.Target, Task.CompletedTask);
```
Unless I'm missing something, it looks like this will still suffer from the same problem that killed the approach attempted in https://github.com/aspnet/Blazor/pull/1651, i.e., that `delegate.Target` won't be what we need if it's a capturing lambda.

Is that something you already have on the list to address for this PR?

The plan we previously came up with to fix this was https://github.com/aspnet/Blazor/pull/1651#issuecomment-438036193. It involves changing the codegen for events and bind.

SteveSandersonMS replied:
OK, I see you're working on addressing this in the commit titled `add event handler infrastructure`.

rynowak replied:
Right. This isn't the fix for that, but these APIs need to change to preserve the existing semantics. These APIs needs to stay put because we still call them in the editor. It's the simplest thing to do to try and keep these APIs working until we completely transition.

SteveSandersonMS commented on src/Components/Components/src/Rendering/ComponentState.cs:
```diff
@@ -97,20 +97,6 @@ public void DisposeInBatch(RenderBatchBuilder batchBuilder)
-            {
-                throw new InvalidOperationException(
-                    $"The component of type {_component.GetType().FullName} cannot receive " +
-                    $"events because it does not implement {typeof(IHandleEvent).FullName}.");
```
**Super** pleased to see this nonsense go away.

SteveSandersonMS commented on src/Components/Components/test/RendererTest.cs:
```diff
@@ -732,46 +731,15 @@ public async Task CanAsyncDispatchEventsToNestedComponents()
         }
 
         [Fact]
-        public async Task ThrowsIfComponentDoesNotHandleEvents()
```
Yay for removing this!

SteveSandersonMS commented on src/Components/Shared/test/TestRenderer.cs:
```diff
@@ -40,12 +40,17 @@ public new Task RenderRootComponentAsync(int componentId)
         public T InstantiateComponent<T>() where T : IComponent
             => (T)InstantiateComponent(typeof(T));
 
+        public Task WaitForAsyncWork()
```
This doesn't seem to be used. Can it be removed?

rynowak replied:
done

SteveSandersonMS commented on src/Components/Components/src/EventHandlerInvokerFactory.cs:
```diff
@@ -0,0 +1,102 @@
+            //
+            // We want to make sure to call StateHasChanged on the Outer component instead
+            // of just the Inner component. We can't rely on action.Target to always point
+            // to the outer component because in the case of a non-capturing lambda it won't.
```
Nit: it's the capturing (not non-capturing) case where delegate.Target won't be `Outer`.

rynowak replied:
I think maybe we're just disagreeing on terminology maybe?

** Capturing Lambda**

```html
<MdButton OnClicked="@(() => DoSomething())" />

@functions {
    void DoSomething() { ... }
}
```


**Non-capturing Lambda**

```html
<MdButton OnClicked="@(() => Console.Log("YEET"))" />
```


**Bound Delegate**

```html
<MdButton OnClicked="DoSomething" />

@functions {
    void DoSomething() { ... }
}
```

SteveSandersonMS replied:
OK yeah. I was distinguishing based on whether the compiler would generate a hidden class to hold captured variables, which it does if and only if you are capturing something *other than* `this`, e.g.:

```html
@foreach(var thing in myCollection)
{
    <MdButton OnClicked="@(() => DoSomething(thing))" />
}
```

In this case, the delegate's `Target` will be an instance of the compiler-generated class instead of the Razor Component.

SteveSandersonMS commented on src/Components/Components/src/EventHandlerInvokerFactory.cs:
```diff
@@ -0,0 +1,102 @@
+                throw new ArgumentNullException(nameof(action));
+            }
+
+            if (receiver is IHandleStateChange handler && !object.ReferenceEquals(receiver, action.Target))
```
Ultra-nit: the most common case will be `receiver == action.Target`, so we could swap the order of these conditionals for an incredibly minor perf gain.

SteveSandersonMS commented on src/Components/Components/test/RendererTest.cs:
```diff
@@ -730,6 +730,160 @@ public async Task CanAsyncDispatchEventsToNestedComponents()
 
+        // The delegate target points to the outer component
+        [Fact]
+        public async Task EventDispatching_DelegateParameter_MethodToDelegateConversion()
```
Is the point of this test to show that it's unnecessary to use `CreateDelegate` if `action.Target` already points to the component you want to be notified after the event?

If so that's great, but maybe add a comment to that effect, as it's pretty subtle!

rynowak replied:
I've added comments to all of these new tests.

SteveSandersonMS commented:
This is all looking extremely positive. Very excited to finally squash the long-standing events weirdnesses, especially as it makes the forms-and-validation scenarios work properly too.

rynowak committed:
Improvements to bind and event handling

The changes here make event dispatching (including bind) more
user-friendly by avoiding the need for manual calls to StateHasChnaged.

We also introduce a new type `EventCallback` (and `EventCallback<T>`).
This is a new primitive that is like a super-powered version of a
delegate. When writing a component that accepts delegates as parameters,
consider using `EventCallback` for the following reasons:
- Allows consumer to pass a variety of different delegate signatures
- Does proper event dispatching and error handling

Using `EventCallback` will eliminate most of the remaining cases where a
manual `StateHasChanged` is required when components are passing content
and delegates to each other.

`EventCallback` is inherently async for the reason that this is really
the only way to provide correct error handling.

-----

The fix for this will be two-phase by first creating a set of APIs that
can be targeted by the compiler that has the desired behaviour and then
updating the compiler to target this new infrastructure.

rynowak commented:
I've pushed an update after incorporating pranav's change but it's not close to done yet. I'm going deep on a few of the specific details of feedback, mostly the async and command ideas. 

One of the major issues to solve is that most of our current stuff is "fire and forget" in nature, and to address that we need to make everything be async.

rynowak commented:
@SteveSandersonMS @javiercn @pranavkm - this is no longer work in progress. Please take a look. meanwhile I'll get busy on the compiler changes, but there's no reason to wait on completing this first since it doesn't really regress things.

rynowak commented on src/Components/Blazor/Build/test/RenderingRazorIntegrationTest.cs:
```diff
@@ -361,26 +362,27 @@ public void SupportsTwoWayBindingForTextboxes()
+
+            // Trigger the change event to show it updates the property
+            await renderer.Invoke(() => setter(new UIChangeEventArgs { Value = "Modified value", }));
+            Assert.Equal("Modified value", myValueProperty.GetValue(component));
```
Now that I've *improved* some of the bind functionality with polyfills these tests require dispatching to the sync context.

SteveSandersonMS replied:
But events are still processed synchronously, aren't they? They absolutely have to be in the WebAssembly case, otherwise we'll kill certain APIs that are only allowed synchronously.

Is is still possible to have unit tests that show event processing happens synchronously?

SteveSandersonMS replied:
Is it OK because in the WebAssembly case, dispatched calls always get executed synchronously?

rynowak replied:
Yes, everything still happens synchronously unless you override `OnHandleEventAsync` and do something async there before calling the delegate.

rynowak commented on src/Components/Browser.JS/src/package-lock.json:
```diff
@@ -1525,7 +1525,8 @@
           "bundled": true,
-          "dev": true
+          "dev": true,
+          "optional": true
```
Will undo this.

rynowak commented on src/Components/Components/src/BindMethods.cs:
```diff
@@ -69,134 +69,200 @@ public static MulticastDelegate GetEventHandlerValue<T>(Func<T, Task> value)
+            return eventArgs =>
+            {
+                setter((bool)((UIChangeEventArgs)eventArgs).Value);
+                _ = DispatchEventAsync(setter.Target, EventCallbackWorkItem.Empty, UIEventArgs.Empty);
```
Explained this in a comment elsewhere but for visibility. Yes, this isn't *super* correct. It's a temporary polyfill until we swap over to the new infrastructure. This is needed before of the change to event dispatching in the renderer. If we don't do something like this we're going to *chicken-or-egg* super hard 🐔

SteveSandersonMS replied:
I haven't completely followed in what sense this isn't correct. Will try to see if the comment elsewhere enlightens me.

rynowak replied:
What is incorrect about is the error handling. It calls `DispatchEventAsync` as fire-and-forget.

rynowak commented on src/Components/Components/src/EventCallbackFactory.cs:
```diff
@@ -0,0 +1,211 @@
+    /// A factory for creating <see cref="EventCallback"/> and <see cref="EventCallback{T}"/>
+    /// instances.
+    /// </summary>
+    public sealed class EventCallbackFactory
```
Instance class to support extension methods. Users will be able to polyfill their own support for `bind` with arbitrary types once compiler work is done.

SteveSandersonMS replied:
Excellent

rynowak commented on src/Components/Components/src/EventCallbackFactoryBinderExtensions.cs:
```diff
@@ -0,0 +1,403 @@
+        /// <param name="setter"></param>
+        /// <param name="existingValue"></param>
+        /// <returns></returns>
+        public static EventCallback<UIChangeEventArgs> CreateBinder(
```
Replacement for old `BindMethods`

rynowak commented on src/Components/Components/src/EventCallbackFactoryUIEventArgsExtensions.cs:
```diff
@@ -0,0 +1,446 @@
+        /// <param name="receiver">The event receiver.</param>
+        /// <param name="callback">The event callback.</param>
+        /// <returns>The <see cref="EventCallback"/>.</returns>
+        public static EventCallback<UIEventArgs> Create(this EventCallbackFactory factory, object receiver, Action<UIEventArgs> callback)
```
These overloads are needed so that method-group-to-delegate conversion will work well. Yeah, it's fun

rynowak commented on src/Components/Components/src/EventCallbackWorkItem.cs:
```diff
@@ -0,0 +1,79 @@
+    /// <summary>
+    /// Wraps a callback delegate associated with an event.
+    /// </summary>
+    public struct EventCallbackWorkItem
```
Suggestions welcome. We want to find an unappealing name for this time so you're likely to use `EventCallback` which is what we want users to find. We could put this in another namespace, but none of the ones we have already are really appropriate. Ultimately we might want to more component lifecycle interfaces elsewhere, and this type would follow.

SteveSandersonMS replied:
I think this name sounds unappealing enough as-is.

rynowak commented on src/Components/Components/src/RenderTree/RenderTreeBuilder.cs:
```diff
@@ -277,6 +277,84 @@ public void AddAttribute(int sequence, string name, MulticastDelegate value)
+            {
+                // If we need to preserve the receiver - we convert this to an untyped EventCallback. We don't
+                // need to preserve the type of an EventCallback<T> when it's invoked from the DOM.
+                Append(RenderTreeFrame.Attribute(sequence, name, (object)value.AsUntyped()));
```
Note: this is a special case. It would be hard to make `RenderTreeDiffBuilder` handle `EventCallback<T>` + DOM events. However we have no such requirement, so I just skip it.

rynowak commented on src/Components/Components/src/Rendering/RendererSynchronizationContext.cs:
```diff
@@ -47,6 +47,10 @@ public Task Invoke(Action action)
+                catch (OperationCanceledException)
+                {
+                    completion.SetCanceled();
+                }
```
This was a bug in the sync context that I found while unit testing related features. We weren't honouring cancellation and would always return a cancelled task as a faulted task instead.

I unwound some of the test infrastructure that had built up and then found this bug. We should be more careful creating "helpers" instead of writing primitive code.

rynowak commented on src/Components/test/testassets/BasicTestApp/EventCallbackTest/EventCallbackCases.cshtml:
```diff
@@ -0,0 +1,36 @@
+    Test cases for using EventCallback with various delegate scenarios that used to be troublesome.
+
+    Currently these cases are **VERBOSE** because we haven't yet landed the compiler support for EventCallback.
+    This will be cleaned up soon, and all of the explicit calls to EventCallback.Factory will go away.
```
Right. So before you complain, I already explained what's going on 😆. Once the compiler work is done all of the rest of this code in the file will become idiomatic.

SteveSandersonMS replied:
Understood!

rynowak commented on src/Components/test/testassets/BasicTestApp/EventCallbackTest/EventCallbackCases.cshtml:
```diff
@@ -0,0 +1,36 @@
+    <TemplatedControl>
+        <button onclick="@EventCallback.Factory.Create(this, () => { })">Unbound Lambda Child Content</button>
+    </TemplatedControl>
+</div>
```
@SteveSandersonMS anything else you think is worth adding?

SteveSandersonMS replied:
When you do `bind-Value` to a component, does the `ValueChanged` invocation now implicitly trigger `StateHasChanged` on the outer component? Not sure whether that is meant to happen automatically given the codegen changes, but if so, would be good to show that too.

rynowak replied:
> When you do `bind-Value` to a component, does the `ValueChanged` invocation now implicitly trigger `StateHasChanged` on the outer component?

Using `Action<T>` will always trigger an event on the delegate target (if possible). We'd recommend using `EventCallback<T>` for this case, which will be supported. 

I'll add a test case for this

rynowak replied:
Note that this requires the component that is producing the bindable to use `EventCallback<T> ValueChanged`.

SteveSandersonMS commented on src/Components/Components/src/BindMethods.cs:
```diff
@@ -69,134 +69,200 @@ public static MulticastDelegate GetEventHandlerValue<T>(Func<T, Task> value)
+            where T : UIEventArgs
+        {
+            return value;
+        }
```
I'm assuming the above two are for type checking in the Razor compiler.

rynowak replied:
exactly.

SteveSandersonMS commented on src/Components/Components/src/ComponentBase.cs:
```diff
@@ -255,9 +255,9 @@ private async Task CallStateHasChangedOnAsyncCompletion(Task task)
+        Task IHandleAfterEvent.HandleEventAsync(EventCallbackWorkItem callback, object arg)
         {
-            var task = binding.Invoke(args);
+            var task = callback.InvokeAsync(arg);
```
Since it actually does invoke the callback itself (instead of just doing stuff *after* the callback was invoked), would the original name `IHandleEvent` actually be a better description?

rynowak replied:
Sure, but this will be my last iteration of renaming this thing until we do an all-up api review.

SteveSandersonMS commented on src/Components/Components/src/EventCallback.cs:
```diff
@@ -0,0 +1,115 @@
+    /// <summary>
+    /// A bound event handler delegate.
+    /// </summary>
+    public readonly struct EventCallback
```
This is such a good feature!

SteveSandersonMS commented on src/Components/Components/src/EventCallbackFactoryBinderExtensions.cs:
```diff
@@ -0,0 +1,403 @@
+            Action<string> setter,
+            string existingValue)
+        {
+            Func<object, string> converter = (obj) => (string)obj;
```
Does this method get called by generated code at each `bind` call site?

If so, instead of instantiating a new `converter` on each call, can we cache the `Func<object, *>` instances on static fields for reuse?

SteveSandersonMS replied:
Here and in the other `CreateBinder` methods, that is.

rynowak replied:
If you'd feel better I can move these to static fields so that it's obvious. 

The compiler already does this optimization where possible. There are different preferences for how deal with this problem when you want to be *sure* that the lambda is static, and they all have drawbacks, other than static local functions which requires C# 8.

SteveSandersonMS commented on src/Components/Components/src/Rendering/Renderer.cs:
```diff
@@ -338,11 +343,28 @@ internal void AssignEventHandlerId(ref RenderTreeFrame frame)
+                // Common cases for this are "chained bind" or "chained event handler" when a component
+                // accepts a delegate as a parameter and then hooks it up to a DOM event.
+                //
+                // When that happens we intentionally box the Command because we need to hold on to
```
In the comment, `Command` -> `EventCallback`?

SteveSandersonMS commented on src/Components/Components/test/EventCallbackTest.cs:
```diff
@@ -0,0 +1,457 @@
+            await callback.InvokeAsync(null);
+        }
+
+
```
Nit: unnecessary blank line here, and in a few other places in this file. I don't really mind whether it gets fixed. I'm kind of only posting this as proof that I'm actually paying attention.

rynowak replied:
Cools beans. I might leave this here, Pranav will be overjoyed if he gets the chance to remove it later.

SteveSandersonMS reviewed (approved):
This looks superb! And a very impressive level of test thoroughness too.

rynowak commented:
Helix has been running now for 2.5 hours - so I'm going to ignore it and just merge this.
